### PR TITLE
AP_RangeFinder: Possible NULL pointer dereference

### PR DESCRIPTION
**Pull request by KimHyungSub**
Issues: #8764
I correct these issues and then open a pull request.
Please check if the code is correct or not.

### CODE REVIEW

**Comment by peterbarker** [comment]
Wow.  This is a doozy - well spotted.

There are several problems here which your PR doesn't address.  Most notably memfd isn't checked for being -1, ctrl for and ram being nullptr.  Additionally, your fix may mean file descriptors are leaked in the case of a failed read.

There are enough resources involved here that I think a `goto CLEANUP` pattern is warranted.  i.e. if an error is detected then you `goto` a cleanup block at the bottom of the function which releases resources, falling through multiple labels as you progressively allocate resources.

(minor cleanup at the same time: don't pre-declare variables at the top, declare when defining, ie.e. `uint32_t memfd = ...`)

Are you happy to chase this more-complete cleanup down?

**Comment by rmackay9** [comment]
@KimHyungSub, it's great that you're finding these issues!  nicely done!

**Comment by KimHyungSub** [comment]
@peterbarker Ah! you are right! I ignore other parts without the *file* descriptor.

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 1]
```diff
@@ -51,14 +51,25 @@ AP_RangeFinder_BBB_PRU::AP_RangeFinder_BBB_PRU(RangeFinder::RangeFinder_State &_
...
-    uint32_t mem_fd;
-    uint32_t *ctrl;
-    void *ram;
-
-    mem_fd = open("/dev/mem", O_RDWR | O_SYNC | O_CLOEXEC);
-    ctrl = (uint32_t*)mmap(0, 0x1000, PROT_READ | PROT_WRITE, MAP_SHARED, mem_fd, PRU0_CTRL_BASE);
-    ram = mmap(0, PRU0_IRAM_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, mem_fd, PRU0_IRAM_BASE);
 
+    uint32_t mem_fd = open("/dev/mem", O_RDWR | O_SYNC | O_CLOEXEC);
+    if(mem_fd == -1)
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: style-guide requires that the brace goes on the same line as the `if`, and a space between the if and the pren.

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 2]
```diff
@@ -51,14 +51,25 @@ AP_RangeFinder_BBB_PRU::AP_RangeFinder_BBB_PRU(RangeFinder::RangeFinder_State &_
...
+    {
+    	return false;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: For consistency, create a label for this case, too.

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 3]
```diff
@@ -51,14 +51,25 @@ AP_RangeFinder_BBB_PRU::AP_RangeFinder_BBB_PRU(RangeFinder::RangeFinder_State &_
...
+	}
+    
+    uint32_t *ctrl = (uint32_t*)mmap(0, 0x1000, PROT_READ | PROT_WRITE, MAP_SHARED, mem_fd, PRU0_CTRL_BASE);
+    if(ctrl == nullptr)
+    {
+    	goto fail_mmap;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: `FAIL_MMAP_CTRL`

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 4]
```diff
@@ -51,14 +51,25 @@ AP_RangeFinder_BBB_PRU::AP_RangeFinder_BBB_PRU(RangeFinder::RangeFinder_State &_
...
+	}
+	
+    void *ram = mmap(0, PRU0_IRAM_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, mem_fd, PRU0_IRAM_BASE);
+    if(ram == nullptr)
+    {
+		goto fail_mmap;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: `FAIL_MMAP_IRAM`

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 5]
```diff
@@ -67,16 +78,16 @@ bool AP_RangeFinder_BBB_PRU::detect()
     FILE *file = fopen("/lib/firmware/rangefinderprutext.bin", "rb");
     if(file == nullptr)
     {
-        result = false;
+        goto fail_fopen;
     }
-
-    if(fread(ram, PRU0_IRAM_SIZE, 1, file) != 1)
+
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: Extraneous whitespace; see if you can turn on "show trailing whitespace" in your editor.

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 6]
```diff
@@ -85,16 +96,17 @@ bool AP_RangeFinder_BBB_PRU::detect()
     file = fopen("/lib/firmware/rangefinderprudata.bin", "rb");
     if(file == nullptr)
     {
-        result = false;
+        goto fail_fopen;
     }
-
-    if(fread(ram, PRU0_DRAM_SIZE, 1, file) != 1)
+    
+	if(fread(ram, PRU0_DRAM_SIZE, 1, file) != 1)
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: Watch out for whitespace changes.  Esp when it's stuffing the indent up :-)

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 7]
```diff
@@ -51,14 +51,25 @@ AP_RangeFinder_BBB_PRU::AP_RangeFinder_BBB_PRU(RangeFinder::RangeFinder_State &_
 bool AP_RangeFinder_BBB_PRU::detect()
 {
     bool result = true;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: Won't need result any more.

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 8]
```diff
@@ -106,7 +118,17 @@ bool AP_RangeFinder_BBB_PRU::detect()
 
     rangerpru = (volatile struct range*)ram;
 
-    return result;
+	return result;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: Can return `true` here... also, whitespace :-)

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 9]
```diff
@@ -50,29 +50,33 @@ AP_RangeFinder_BBB_PRU::AP_RangeFinder_BBB_PRU(RangeFinder::RangeFinder_State &_
...
-    bool result = true;
-    uint32_t mem_fd;
-    uint32_t *ctrl;
-    void *ram;
+    uint32_t mem_fd = open("/dev/mem", O_RDWR | O_SYNC | O_CLOEXEC);
+    if(mem_fd == -1) {
+        goto FAIL_MEM_FD;
+    }
+
+    uint32_t *ctrl = (uint32_t*)mmap(0, 0x1000, PROT_READ | PROT_WRITE, MAP_SHARED, mem_fd, PRU0_CTRL_BASE);
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: This must be unmapped in the relevant failure cases.

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 10]
```diff
@@ -50,29 +50,33 @@ AP_RangeFinder_BBB_PRU::AP_RangeFinder_BBB_PRU(RangeFinder::RangeFinder_State &_
 */
 bool AP_RangeFinder_BBB_PRU::detect()
 {
-    bool result = true;
-    uint32_t mem_fd;
-    uint32_t *ctrl;
-    void *ram;
+    uint32_t mem_fd = open("/dev/mem", O_RDWR | O_SYNC | O_CLOEXEC);
+    if(mem_fd == -1) {
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: spaces after `if`s

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 11]
```diff
@@ -83,14 +87,12 @@ bool AP_RangeFinder_BBB_PRU::detect()
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: return value of this mmap isn't checked

File: libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp [hunk 12]
```diff
@@ -83,14 +87,12 @@ bool AP_RangeFinder_BBB_PRU::detect()
 
     // Load firmware (.data)
     file = fopen("/lib/firmware/rangefinderprudata.bin", "rb");
-    if(file == nullptr)
-    {
-        result = false;
+    if(file == nullptr) {
+        goto FAIL_FOPEN;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BBB_PRU.cpp]: You will need a separate label for this. `FAIL_FOPEN_DATA`, perhaps?

**Comment by khancyr** [comment]
This is bitrooting and low on safety as it touch only the PRU code. I will close this. Feel free to reopen with the requested changes. thanks anyway

**Pull request closed without merging**